addons/point_of_sale/models/pos_order.py

```diff
@@ -1420,51 +1422,161 @@ def get_sale_details(self, date_start=False, date_stop=False, config_ids=False,
-                    for tax in line_taxes['taxes']:
-                        taxes.setdefault(tax['id'], {'name': tax['name'], 'tax_amount':0.0, 'base_amount':0.0})
-                        taxes[tax['id']]['tax_amount'] += tax['amount']
-                        taxes[tax['id']]['base_amount'] += tax['base']
+                if line.qty >= 0:
+                    key1 = line.product_id.product_tmpl_id.pos_categ_id.name
+                    key2 = (line.product_id, line.price_unit, line.discount)
+                    products_sold.setdefault(key1, {})
+                    products_sold[key1].setdefault(key2, 0.0)
+                    products_sold[key1][key2] += line.qty
+
+                    if line.tax_ids_after_fiscal_position:
+                        line_taxes = line.tax_ids_after_fiscal_position.sudo().compute_all(line.price_unit * (1-(line.discount or 0.0)/100.0), currency, line.qty, product=line.product_id, partner=line.order_id.partner_id or False)
+                        for tax in line_taxes['taxes']:
+                            taxes.setdefault(tax['id'], {'name': tax['name'], 'tax_amount':0.0, 'base_amount':0.0})
+                            taxes[tax['id']]['tax_amount'] += tax['amount']
+                            taxes[tax['id']]['base_amount'] += tax['base']
+                    else:
+                        taxes.setdefault(0, {'name': _('No Taxes'), 'tax_amount':0.0, 'base_amount':0.0})
+                        taxes[0]['base_amount'] += line.price_subtotal_incl
                 else:
-                    taxes.setdefault(0, {'name': _('No Taxes'), 'tax_amount':0.0, 'base_amount':0.0})
-                    taxes[0]['base_amount'] += line.price_subtotal_incl
+                    key1 = line.product_id.product_tmpl_id.pos_categ_id.name
+                    key2 = (line.product_id, line.price_unit, line.discount)
+                    refund_done.setdefault(key1, {})
+                    refund_done[key1].setdefault(key2, 0.0)
+                    refund_done[key1][key2] += line.qty
+
+                    if line.tax_ids_after_fiscal_position:
+                        line_taxes = line.tax_ids_after_fiscal_position.sudo().compute_all(line.price_unit * (1-(line.discount or 0.0)/100.0), currency, line.qty, product=line.product_id, partner=line.order_id.partner_id or False)
+                        for tax in line_taxes['taxes']:
+                            refund_taxes.setdefault(tax['id'], {'name': tax['name'], 'tax_amount':0.0, 'base_amount':0.0})
+                            refund_taxes[tax['id']]['tax_amount'] += tax['amount']
+                            refund_taxes[tax['id']]['base_amount'] += tax['base']
+                    else:
+                        refund_taxes.setdefault(0, {'name': _('No Taxes'), 'tax_amount':0.0, 'base_amount':0.0})
+                        refund_taxes[0]['base_amount'] += line.price_subtotal_incl
```

Comment:
This is a complete repeat of the code above. Instead of duplicating the code by using another variable, can you refactor it to make it cleaner? you can create new method for such case